lib/covalence/core/state_stores/s3.rb

```diff
@@ -34,12 +34,16 @@ def get_doc(bucket, document)
         @cache[bucket][document] ||= begin
           @s3.get_object(bucket: bucket, key: document).body.read
         rescue Aws::S3::Errors::ServiceError => err
-          fail "Unable to retrieve document '#{document}' from bucket '#{bucket}': " + err.message
+          raise err, err.message + " Unable to retrieve document '#{document}' from bucket '#{bucket}'"
         end
       end
 
       def get_key(bucket, document, name)
-        doc = self.get_doc(bucket, document)
+        begin
+          doc = self.get_doc(bucket, document)
+        rescue Aws::S3::Errors::ServiceError => err
+          fail err.message + " and the key is: #{name}. Check your variables in .yaml files."
```

Comment:
What do you think about `" for key '#{name}'. Check your data files."` for conformity with the first part of the message? It's a minor nit, but Hiera can technically support backends other than YAML.